snafu-derive/src/lib.rs

```diff
@@ -451,31 +471,34 @@ fn parse_snafu_enum(
                 let span = syn_field.span();
                 let name = syn_field
                     .ident
+                    .as_ref()
                     .ok_or_else(|| vec![syn::Error::new(span, "Must have a named field")])?;
                 let field = Field {
                     name: name.clone(),
-                    ty: syn_field.ty,
+                    ty: syn_field.ty.clone(),
                 };
 
-                let mut has_backtrace_delegate = false;
-                let mut is_backtrace = None;
-                let mut is_source = None;
-                let mut transformations =
-                    AtMostOne::new("source(from)", "within an error variant");
-
                 // Check whether we have multiple source/backtrace attributes on this field.
                 // We can't just add to source_fields/backtrace_fields from inside the attribute
                 // loop because source and backtrace are connected and require a bit of special
                 // logic after the attribute loop.  For example, we need to know whether there's a
                 // source transformation before we record a source field, but it might be on a
                 // later attribute.  We don't actually need any data out of this, it's just a
                 // convenient way to check for dupes and get nice errors.
-                // Note: if there are more than two, the errors generated are identical and dedupe
-                // down to one error; it should be enough to find the problem.
                 let mut source_attrs = AtMostOne::new("source", "on one field");
+                let mut transformations = AtMostOne::new("source(from)", "on one field");
                 let mut backtrace_attrs = AtMostOne::new("backtrace", "on one field");
 
-                for attr in attributes_from_syn(syn_field.attrs)? {
+                // Keep track of the negative markers so we can check for inconsistencies and
+                // exclude fields even if they have the "source" or "backtrace" name.
+                let mut has_source_false = false;
+                let mut has_backtrace_false = false;
```

Comment:
Ooof. Double negatives make my head hurt. I don't have the brainpower right now to think of an alternative, but I'd like to see if there's something we can come up with.

Reply:
The original `is_source`, `is_backtrace`, and `is_backtrace_delegate` were certainly clear in their intention, but you had to be (IMHO) more careful about when to flip and check those.  Using the existing `AtMostOne`s felt natural since they were already tracking which attributes we'd seen, and it eliminated duplicated state.  That just left the "odd" case of a user denying that the field is what it appears to be.  An alternative could be to add a method to `AtMostOne` that lets you override what it's seen, guaranteeing it will only give errors upon `finish`.  That doesn't feel particularly clean to me either, though.  Hmm.

Reply:
What about using something like "opt out"? Maybe like `source_opt_out = false`?